Various improvements: create log, send email, config file, include/exclude list of subjects

cfhammill opened the pull request:
Implementing the following features/improvements to `sct_run_batch.py`:

- Display header "Spinal Cord Toolbox (version)" when running the command, as done in other SCT functions;
- Create log file (new flag: `-batch-log`, with defaul: sct_run_batch_log.txt), of whole Terminal output;
- Added an `-email-to` feature to notify when processing is finished;
- Possible to include a json or yaml config file that replaces the input arguments, with flag `-config`;

Fixes #2726

cfhammill commented:
I just pushed a pretty large commit that adds the ability to pass json and yaml configuration. Some key design choices that might be worth discussing

- I didn't use a pre-canned solution for configuration parsing. I rolled a very small wrapper around reading json/yml and updating the argparse results. I am aware of ConfigArgParse, but it seemed more heavy weight and I couldn't find a clear enough description of how it handles precedence. 
- For precedence I chose: command line arguments overwrite configuration options. and issues a warning saying that using this feature is discouraged. I figure this shouldn't be an error, when you're trying things you may wish to temporarily overwrite the configuration, but I'm open to making this an error.
- I implemented the precedence by running two passes of the command line parser. Parse once to create a dictionary of acceptable arguments. Then read in the config file, checking for unrecognized keys against the acceptable argument dictionary. Then use the parsers `update_defaults` method to set the config file options as the new defaults. Then reparse the command line.
- Unrecognized keys in the config are presently a warning, but I think this could potentially be an error as well.
- To handle IO redirection, I implemented a small `Tee` class in `sct_utils`. There are some difficulties with this. Including a proper `__del__` method breaks pytest's output capture somehow, and without it I need to manually close the log at the end of the script. This would probably be better handled by our logging infrastructure, but I couldn't figure out how to do it simply. Maybe this can be opened as an issue for myself or someone to address in the future.
- I added tests for a few of the key errors and warnings to `unit_testing`, but it would be nice to have a few tests that check that a real task file could be run. This wouldn't be too hard add to the test suite but haven't got there yet. The tests aren't nearly exhaustive so I would appreciate if someone could try it out and report back.

Still to do:

- Adding notifications
- Printing command line args and configuration options. Probably will just print the internal dictionary of the argparse `Namespace` after the second parse pass.

jcohenadad commented:
> I just pushed a pretty large commit that adds the ability to pass json and yaml configuration. Some key design choices that might be worth discussing
> 
> * I didn't use a pre-canned solution for configuration parsing. I rolled a very small wrapper around reading json/yml and updating the argparse results. I am aware of ConfigArgParse, but it seemed more heavy weight and I couldn't find a clear enough description of how it handles precedence.

i'm fine with your decision

> * For precedence I chose: command line arguments overwrite configuration options. and issues a warning saying that using this feature is discouraged. I figure this shouldn't be an error, when you're trying things you may wish to temporarily overwrite the configuration, but I'm open to making this an error.

i'm not sure i understand the context here-- are you referring to the situation where someone would point to a json/yml file, and at the same time use a parser flag (e.g. -path-output) that conflicts with the input file?  If that's what you meant, then i agree that issuing an error is not desired here. Simple warning should suffice. 

> * I implemented the precedence by running two passes of the command line parser. Parse once to create a dictionary of acceptable arguments. Then read in the config file, checking for unrecognized keys against the acceptable argument dictionary. Then use the parsers `update_defaults` method to set the config file options as the new defaults. Then reparse the command line.

that sounds reasonable

> * Unrecognized keys in the config are presently a warning, but I think this could potentially be an error as well.

i'm fine with warning (as long as they also appear in the output log file)

> * To handle IO redirection, I implemented a small `Tee` class in `sct_utils`. There are some difficulties with this. Including a proper `__del__` method breaks pytest's output capture somehow, and without it I need to manually close the log at the end of the script. This would probably be better handled by our logging infrastructure, but I couldn't figure out how to do it simply. Maybe this can be opened as an issue for myself or someone to address in the future.

we are trying to (slowly) move `sct_utils` to `spinalcordtoolbox.utils`, so for any new functions we encourage implementing them directly in `spinalcordtoolbox.utils`. Regarding the issue you are referring to, i don't have enough knowledge for giving you feedback. @kousu @Drulex @bcdarwin?

> * I added tests for a few of the key errors and warnings to `unit_testing`, but it would be nice to have a few tests that check that a real task file could be run. This wouldn't be too hard add to the test suite but haven't got there yet. The tests aren't nearly exhaustive so I would appreciate if someone could try it out and report back.

i'll try, but i don't guaranty i will cover all funky scenarios

> 
> Still to do:
> 
> * Adding notifications
> * Printing command line args and configuration options. Probably will just print the internal dictionary of the argparse `Namespace` after the second parse pass.

👍

cfhammill commented:
> > * For precedence I chose: command line arguments overwrite configuration options. and issues a warning saying that using this feature is discouraged. I figure this shouldn't be an error, when you're trying things you may wish to temporarily overwrite the configuration, but I'm open to making this an error.
> 
> i'm not sure i understand the context here-- are you referring to the situation where someone would point to a json/yml file, and at the same time use a parser flag (e.g. -path-output) that conflicts with the input file? If that's what you meant, then i agree that issuing an error is not desired here. Simple warning should suffice.
> 

yes, that's the context I mean.

> > * Unrecognized keys in the config are presently a warning, but I think this could potentially be an error as well.
> 
> i'm fine with warning (as long as they also appear in the output log file)

they should given the `Tee` class below is working properly

> > * To handle IO redirection, I implemented a small `Tee` class in `sct_utils`. There are some difficulties with this. Including a proper `__del__` method breaks pytest's output capture somehow, and without it I need to manually close the log at the end of the script. This would probably be better handled by our logging infrastructure, but I couldn't figure out how to do it simply. Maybe this can be opened as an issue for myself or someone to address in the future.
> 
> we are trying to (slowly) move `sct_utils` to `spinalcordtoolbox.utils`, so for any new functions we encourage implementing them directly in `spinalcordtoolbox.utils`. Regarding the issue you are referring to, i don't have enough knowledge for giving you feedback. @kousu @Drulex @bcdarwin?
> 

Sorry this is in fact in `spinalcordtoolbox.utils` pardon me. Hopefully one of the above is willing to review.

> > * I added tests for a few of the key errors and warnings to `unit_testing`, but it would be nice to have a few tests that check that a real task file could be run. This wouldn't be too hard add to the test suite but haven't got there yet. The tests aren't nearly exhaustive so I would appreciate if someone could try it out and report back.
> 
> i'll try, but i don't guaranty i will cover all funky scenarios
> 

Thanks, I'll be slowly expanding the set of test cases as issues crop up, trying to catch all the odd corner cases.

bcdarwin commented:
> > > * To handle IO redirection, I implemented a small `Tee` class in `sct_utils`. There are some difficulties with this. Including a proper `__del__` method breaks pytest's output capture somehow, and without it I need to manually close the log at the end of the script. This would probably be better handled by our logging infrastructure, but I couldn't figure out how to do it simply. Maybe this can be opened as an issue for myself or someone to address in the future.
> > 
> > 
> > we are trying to (slowly) move `sct_utils` to `spinalcordtoolbox.utils`, so for any new functions we encourage implementing them directly in `spinalcordtoolbox.utils`. Regarding the issue you are referring to, i don't have enough knowledge for giving you feedback. @kousu @Drulex @bcdarwin?
> 
> Sorry this is in fact in `spinalcordtoolbox.utils` pardon me. Hopefully one of the above is willing to review.

You haven't posted the error which arises, but could this be related to [SIGPIPE handling](https://stackoverflow.com/questions/11423225/why-does-my-python3-script-balk-at-piping-its-output-to-head-or-tail-sys-module)?

cfhammill commented:
@bcdarwin to reproduce this issue, just uncomment the `__del__` method of `Tee` in `spinalcordtoolbox.utils`. Then run `pytest unit_testing/test_sct_run_batch.py`
[tee-io-fails.txt](https://github.com/neuropoly/spinalcordtoolbox/files/4719855/tee-io-fails.txt)

The relevant error is:

```
======================================================================================================= FAILURES ========================================================================================================
_________________________________________________________________________________________________ test_only_one_include _________________________________________________________________________________________________

self = <contextlib._GeneratorContextManager object at 0x7f84ffffdcf8>, type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):
        if type is None:
            try:
>               next(self.gen)
E               ValueError: I/O operation on closed file

python/envs/venv_sct/lib/python3.6/contextlib.py:88: ValueError
================================================================================================ short test summary info ================================================================================================
FAILED unit_testing/test_sct_run_batch.py::test_only_one_include - ValueError: I/O operation on closed file
ERROR unit_testing/test_sct_run_batch.py::test_only_one_include - OSError: [Errno 9] Bad file descriptor
========================================================================================= 1 failed, 2 passed, 1 error in 0.49s ==========================================================================================
Traceback (most recent call last):
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/bin/pytest", line 8, in <module>
    sys.exit(main())
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/config/__init__.py", line 125, in main
    config=config
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/pluggy/manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/main.py", line 240, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/main.py", line 235, in wrap_session
    config._ensure_unconfigure()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/config/__init__.py", line 829, in _ensure_unconfigure
    fin()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/capture.py", line 129, in stop_global_capturing
    self._global_capturing.pop_outerr_to_orig()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/capture.py", line 474, in pop_outerr_to_orig
    out, err = self.readouterr()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/capture.py", line 515, in readouterr
    out = self.out.snap()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/capture.py", line 626, in snap
    res = super().snap()
  File "/home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/_pytest/capture.py", line 585, in snap
    self.tmpfile.seek(0)
ValueError: I/O operation on closed file
```

cfhammill commented:
After discussing at the meeting the consensus is that very few users are going to be using the email feature so it is reasonable to ask them to configure email sending from their machine. This eliminates the need for using passwords and tokens for third party software. Conveniently for me this reduces the implementation burden.

jcohenadad commented:
@cfhammill what is the status of this PR?

jcohenadad commented:
@cfhammill after spending 30min finding why my `-seg-manual` flag was not considered, i realized that it fails when feeding a relative path (because later on, current location changes). So i would suggest to immediately change every input paths (for seg-manual, path-output, etc.) into absolute path, to avoid that problem.

cfhammill commented:
just coming back from from some time away @jcohenadad, this PR is pretty close I think. Blocking right now is the email feature. I prototyped using opensmtp as suggested by @kousu, with a little bit of work it works great. 

I think `path-segmanual` was the last path that wasn't made absolute, but I'll ensure all paths are.

cfhammill commented:
I hit a bit of a snag with the email notifications, it looks like that while running an `opensmtpd` server on my machine, I ran in to a situation where the emails weren't delivered for over 12 hours. It seems like this solution would be unreliable in practice. Further we need a guarantee that the password is not prompted when the email is to be sent (e.g. the msmtp solution proposed by @zougloub), because the user would need to be around when the email gets sent.

Given that this is primarily for internal use, I'm going to take the email interface from `sct_pipeline` and adapt it so that it prompts for the password using `getpass` at the start of the script.

cfhammill commented:
@jcohenadad, email feature is ready, `path_segmanual` is absolutized, and I logged the argument/config via yaml.

Ready for a review whenever someone wants to take that on.

cfhammill commented:
Remaining TODOs perhaps for a different PR:
- argument groups for argparse might clean the interface up a bit
- investigate the `Tee` class to make sure it does what it's supposed to
- add more tests, I added a handful of trivial interface tests, but it might be worth doing this more completely (this may be worth doing for this PR)

jcohenadad commented:
Running into problem when using `-include-list`:

~~~
ulien-macbook:~/temp/test_chris_batch $ sct_run_batch -path-data /Users/julien/data/ds002900 -include-list sub-perform ~/code/spine-generic/processing/process_data.sh 

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-102284751c8ea200b8ca9850a5952f165337d0e8)

usage: sct_run_batch [-h] [-config CONFIG] [-jobs <int>] [-path-data PATH_DATA]
                     [-subject-prefix SUBJECT_PREFIX] [-path-output PATH_OUTPUT]
                     [-batch-log BATCH_LOG] [-include INCLUDE]
                     [-include-list INCLUDE_LIST [INCLUDE_LIST ...]] [-exclude EXCLUDE]
                     [-exclude-list EXCLUDE_LIST [EXCLUDE_LIST ...]]
                     [-path-segmanual PATH_SEGMANUAL] [-itk-threads <int>] [-task-args TASK_ARGS]
                     [-email-to EMAIL_TO] [-email-from EMAIL_FROM] [-email-host EMAIL_HOST]
                     task
sct_run_batch: error: the following arguments are required: task
~~~

jcohenadad commented:
@cfhammill did you try the email sending feature with gmail's smtp? they blocked it, sent me an email about unusual activity, and after i clicked on "that was me", they sent me this: 
<img width="526" alt="Screen Shot 2020-06-25 at 3 34 48 PM" src="https://user-images.githubusercontent.com/2482071/85787537-8b653880-b6f9-11ea-9f6f-fc79036177bb.png">

jcohenadad commented:
Tried with -config flag and running into this:

<details><summary>syntax+error</summary>

~~~
julien-macbook:~/temp/test_chris_batch $ sct_run_batch -config config.yml ~/code/spine-generic/processing/process_data.sh 

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-102284751c8ea200b8ca9850a5952f165337d0e8)

configuring
/Users/julien/code/sct/scripts/sct_run_batch.py:173: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config = yaml.load(conf)
/Users/julien/code/sct/scripts/sct_run_batch.py:177: UserWarning: Using the `-config|-c` flag with additional arguments is discouraged
  warnings.warn(UserWarning('Using the `-config|-c` flag with additional arguments is discouraged'))
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 348, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 181, in main
    config_keys = set(config.keys())
AttributeError: 'str' object has no attribute 'keys'
~~~

</details>

i am using this YAML file:
```yaml
 path_data /Users/julien/data/_test_julien/
```

jcohenadad commented:
also, we should probably refer to example json and yaml in the documentation

jcohenadad commented:
Another issue when trying with json:

<details><summary>syntax+error</summary>

~~~
julien-macbook:~/temp/test_chris_batch $ sct_run_batch -config config.json ~/code/spine-generic/processing/process_data.sh 

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-102284751c8ea200b8ca9850a5952f165337d0e8)

configuring
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 348, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 171, in main
    config = json.load(conf)
  File "/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/json/__init__.py", line 299, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 2 column 3 (char 4)
~~~

</details>

my json file:
```xml
{
  path_data: "/Users/julien/data/_test_julien/"
}
```

cfhammill commented:
Thanks for kicking the tires @jcohenadad!

For the less secure apps, yes I enabled this for my account. It's annoying. It would be nice to figure out a token system. I can investigate if you think this is worth fixing for this PR.

For the two config files the yaml variant was just missing a colon

```yml
path_data: /Users/julien/data/_test_julien/
```

The json version was missing quotes on the property name

```js
{
  "path_data" : "/Users/julien/data/_test_julien/"
}
```

but I agree we should include examples in the documentation to avoid users getting tripped by syntax.

The parsing failure on `-include-list` is more troubling. I'll investigate what went wrong there.

cfhammill commented:
I guess this is known `argparse` behaviour https://stackoverflow.com/questions/26985650/argparse-do-not-catch-positional-arguments-with-nargs

It can be fixed pretty simply (but not necessarily intuitively by adding a `--` spacer)

```bash
sct_run_batch -path-data /Users/julien/data/ds002900 -include-list sub-perform -- ~/code/spine-generic/processing/process_data.sh
```

or reordering the arguments so that the list argument doesn't come last.

We could move `task` from a positional to a named argument at the expense of slightly more typing? Or I since we're already manually munging the commandline args, we could try parsing the args catching this case. Inserting the spacer before the last argument if we notice this problem.

jcohenadad commented:
> For the less secure apps, yes I enabled this for my account. It's annoying. It would be nice to figure out a token system. I can investigate if you think this is worth fixing for this PR.

no worries, let's do it in another PR. 

> but I agree we should include examples in the documentation to avoid users getting tripped by syntax.

yup! i could have spent 5min reading about json/yaml convention but i also like to play the dummy user for this reason 😉 

p.s. i've re-tried with the proper syntax and it works well with both json and yaml 👍 

> We could move task from a positional to a named argument at the expense of slightly more typing? 

i would definitely suggest that. Every time i use the function, i have to remember that task is a positional argument (not common in SCT ecosystem)

cfhammill commented:
Great, I've switched task to named and added example JSON and YAML config for syntax references

jcohenadad commented:
@cfhammill what is the status of this PR? is it soon ready to be merged? thanks

jcohenadad commented:
running into this bug with 6ee10d69e32f7969d2337cf6ce4845b234af529f using this config file: 
```yaml
path_data: /Users/julien/code/spine-generic/data-single-subject
path_output: /Users/julien/data/spine_generic_single_20200703
task: /Users/julien/code/spine-generic/spine-generic/processing/process_data.sh
include: douglas
jobs: 6
```

~~~
julien-macbook:~/data/spine_generic_single_20200703 $ sct_run_batch -c config.yml 

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-6ee10d69e32f7969d2337cf6ce4845b234af529f)

configuring
/Users/julien/code/sct/scripts/sct_run_batch.py:188: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config = yaml.load(conf)
/Users/julien/code/sct/scripts/sct_run_batch.py:192: UserWarning: Using the `-config|-c` flag with additional arguments is discouraged
  warnings.warn(UserWarning('Using the `-config|-c` flag with additional arguments is discouraged'))
sct_run_batch arguments were:
batch_log: sct_run_batch_log.txt
config: config.yml
email_from: null
email_host: smtp.gmail.com:587
email_to: null
exclude: null
exclude_list: null
include: douglas
include_list: null
itk_threads: 1
jobs: 6
path_data: /Users/julien/code/spine-generic/data-single-subject
path_output: /Users/julien/data/spine_generic_single_20200703
path_segmanual: .
subject_prefix: sub-
task: /Users/julien/code/spine-generic/spine-generic/processing/process_data.sh
task_args: ''

Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 363, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 283, in main
    'Only one of `include` and `include-list` can be used'
AssertionError: Only one of `include` and `include-list` can be used
~~~

jcohenadad commented:
https://github.com/neuropoly/spinalcordtoolbox/pull/2732#issuecomment-653600981 fixed in 5790a7d

cfhammill commented:
Thanks for the fixes @jcohenadad! I think this PR is pretty close. I added a boolean flag to configure the fail fast behaviour. I haven't had a chance to make sure it works but I'll put it up anyway in case you want to give it a try. Otherwise I'll try to stress test it tomorrow.

jcohenadad commented on scripts/sct_run_batch.py:
```diff
@@ -21,146 +21,365 @@
+
+    if do_email:
+        send_notification('sct_run_batch: Run completed',
+                          fails_message + completed_message)
```
`fails_message` is not declared. Did you mean `fails`?

cfhammill replied:
whoops, that should be `status_message`, fixed

jcohenadad commented:
i am getting the following error (a922cd4):
~~~
julien-macbook:~/data/spine_generic_single_20200703/test $ sct_run_batch -c config.yml 

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-a922cd440a18eca8bc99fb9998cc1f6cb974f043)

configuring
/Users/julien/code/sct/scripts/sct_run_batch.py:199: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config = yaml.load(conf)
/Users/julien/code/sct/scripts/sct_run_batch.py:203: UserWarning: Using the `-config|-c` flag with additional arguments is discouraged
  warnings.warn(UserWarning('Using the `-config|-c` flag with additional arguments is discouraged'))
sct_run_batch arguments were:
batch_log: sct_run_batch_log.txt
config: config.yml
continue_on_error: 1
email_from: null
email_host: smtp.gmail.com:587
email_to: null
exclude: null
exclude_list: null
include: douglas
include_list: null
itk_threads: 1
jobs: 6
path_data: /Users/julien/code/spine-generic/data-single-subject
path_output: .
path_segmanual: .
subject_prefix: sub-
task: /Users/julien/code/spine-generic/spine-generic/processing/process_data.sh
task_args: ''

Running sub-douglas. See log file /Users/julien/data/spine_generic_single_20200703/test/log/process_data_sub-douglas.log
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 385, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 355, in main
    fails = [sd for (sd, ret) in zip(subject_dirs, results) if ret.returncode != 1]
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 355, in <listcomp>
    fails = [sd for (sd, ret) in zip(subject_dirs, results) if ret.returncode != 1]
AttributeError: 'NoneType' object has no attribute 'returncode'
~~~

jcohenadad commented:
@cfhammill with https://github.com/neuropoly/spinalcordtoolbox/pull/2732/commits/cf97c1b80838570e99cd675b43fbd0f959e7776c, `sct_run_batch` does continue until the end, but i still get the following error:
~~~
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/jcohen/code/sct/python/envs/venv_sct/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/jcohen/code/sct/python/envs/venv_sct/lib/python3.6/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/home/jcohen/code/sct/scripts/sct_run_batch.py", line 169, in run_single
    assert res.returncode == 0, 'Processing of subject {} failed'.format(subject)
AssertionError: Processing of subject sub-tokyoSigna2 failed
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jcohen/code/sct/scripts/sct_run_batch.py", line 363, in <module>
    main(sys.argv[1:])
  File "/home/jcohen/code/sct/scripts/sct_run_batch.py", line 334, in main
    raise e
  File "/home/jcohen/code/sct/scripts/sct_run_batch.py", line 323, in main
    p.map(run_single_dir, subject_dirs)
  File "/home/jcohen/code/sct/python/envs/venv_sct/lib/python3.6/multiprocessing/pool.py", line 266, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/home/jcohen/code/sct/python/envs/venv_sct/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
AssertionError: Processing of subject sub-tokyoSigna2 failed
~~~

which is annoying because i would like to know how long the whole analysis lasted. Also, this is not an elegant way to end the function. Instead of crashing the function, why not continuing `sct_run_batch` until the end, without raising any script-related error?   if there is an error in the script, it should be easily identified by the `err.` prefix. What do you thing?

cfhammill commented:
@jcohenadad, clearly something isn't working, it shouldn't continue to the end and save that for the logs. I'll investigate why my modifications didn't have the desired effect.

cfhammill commented:
@jcohenadad Can you try to reproduce the error you saw in your previous comment? That error looks like the behaviour from before these modification, that error should get trapped (it is unrelated to the missing return noted above).

cfhammill commented:
Nevermind, there was an error in the local variables lookup which caused the error to bubble up

cfhammill commented:
Ok, should be ready for another kick @jcohenadad, sorry for the inconvenience.

jcohenadad commented:
Edge case, but which could happen, is when specifying output folder with numerical values, e.g.:

```yaml
path_data: /Users/julien/code/spine-generic/data-single-subject
path_output: 20200708142245
task: /Users/julien/code/spine-generic/spine-generic/processing/process_data.sh
jobs: 6
include: douglas
```

Gives me this error:
~~~
julien-macbook:~/data/spine_generic_single_20200703 $ sct_run_batch -c config.yml 

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-926347ba2b5ec7db45d1df427cf0f0bba9b0acb6)

configuring
/Users/julien/code/sct/scripts/sct_run_batch.py:201: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config = yaml.load(conf)
/Users/julien/code/sct/scripts/sct_run_batch.py:205: UserWarning: Using the `-config|-c` flag with additional arguments is discouraged
  warnings.warn(UserWarning('Using the `-config|-c` flag with additional arguments is discouraged'))
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 388, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/scripts/sct_run_batch.py", line 254, in main
    path_output = os.path.abspath(os.path.expanduser(args.path_output))
  File "/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/posixpath.py", line 235, in expanduser
    path = os.fspath(path)
TypeError: expected str, bytes or os.PathLike object, not int
~~~

jcohenadad commented:
After running a process, a message states: _"Your batch completed but some subjects may have not completed successfully, please consult the logs for:"_, and then lists all the logs that successfully ended, whereas i would expect to see a list of failed logs instead. 

<details><summary>Details</summary>

~~~
[jcohen@cedar5 log]$ cat sct_run_batch_log.txt

--
Spinal Cord Toolbox (git-cfh/sct-run-batch-improvements_2726-926347ba2b5ec7db45d1df427cf0f0bba9b0acb6)

sct_run_batch arguments were:
batch_log: sct_run_batch_log.txt
config: /home/jcohen/config.yml
continue_on_error: 1
email_from: null
email_host: smtp.gmail.com:587
email_to: null
exclude: null
exclude_list: null
include: null
include_list: null
itk_threads: 1
jobs: 0
path_data: /scratch/jcohen/data-single-subject
path_output: /scratch/jcohen/results
path_segmanual: .
subject_prefix: sub-
task: /home/jcohen/code/spine-generic/spine-generic/processing/process_data.sh
task_args: ''

Running sub-glen. See log file /scratch/jcohen/results/log/process_data_sub-glen.log
Running sub-juntendoPrisma. See log file /scratch/jcohen/results/log/process_data_sub-juntendoPrisma.log
Running sub-tokyo750w. See log file /scratch/jcohen/results/log/process_data_sub-tokyo750w.log
Running sub-ucl. See log file /scratch/jcohen/results/log/process_data_sub-ucl.log
Running sub-juntendoAchieva. See log file /scratch/jcohen/results/log/process_data_sub-juntendoAchieva.log
Running sub-juntendoSkyra. See log file /scratch/jcohen/results/log/process_data_sub-juntendoSkyra.log
Running sub-oxfordFmrib. See log file /scratch/jcohen/results/log/process_data_sub-oxfordFmrib.log
Running sub-tokyoSigna1. See log file /scratch/jcohen/results/log/process_data_sub-tokyoSigna1.log
Running sub-unf. See log file /scratch/jcohen/results/log/process_data_sub-unf.log
Running sub-mgh. See log file /scratch/jcohen/results/log/process_data_sub-mgh.log
Running sub-tokyoIngenia. See log file /scratch/jcohen/results/log/process_data_sub-tokyoIngenia.log
Running sub-chiba750. See log file /scratch/jcohen/results/log/process_data_sub-chiba750.log
Running sub-perform. See log file /scratch/jcohen/results/log/process_data_sub-perform.log
Running sub-chibaIngenia. See log file /scratch/jcohen/results/log/process_data_sub-chibaIngenia.log
Running sub-poly. See log file /scratch/jcohen/results/log/process_data_sub-poly.log
Running sub-douglas. See log file /scratch/jcohen/results/log/process_data_sub-douglas.log
Running sub-juntendo750w. See log file /scratch/jcohen/results/log/process_data_sub-juntendo750w.log
Running sub-tokyoSigna2. See log file /scratch/jcohen/results/log/process_data_sub-tokyoSigna2.log
Running sub-tokyoSkyra. See log file /scratch/jcohen/results/log/process_data_sub-tokyoSkyra.log
Your batch completed but some subjects may have not completed successfully, please consult the logs for:
sub-glen
sub-oxfordFmrib
sub-ucl
sub-juntendoSkyra
sub-juntendoAchieva
sub-tokyo750w
sub-juntendoPrisma
sub-unf
sub-mgh
sub-tokyoSigna1
sub-tokyoIngenia
sub-perform
sub-chiba750
sub-chibaIngenia
sub-poly
sub-douglas
sub-juntendo750w
sub-tokyoSkyra
Finished 
Started: 13:08
Ended: 13:38

To open the Quality Control (QC) report on a web-browser, run the following:
xdg-open /scratch/jcohen/results/qc/index.html
[jcohen@cedar5 log]$ ll 
total 4580
-rw-r----- 1 jcohen jcohen    206 Jul  8 13:26 _error_check_output_files.log
-rw-r----- 1 jcohen jcohen 245188 Jul  8 13:32 err.process_data_sub-tokyoSigna2.log
-rw-r----- 1 jcohen jcohen 250581 Jul  8 13:38 process_data_sub-chiba750.log
-rw-r----- 1 jcohen jcohen 245650 Jul  8 13:33 process_data_sub-chibaIngenia.log
-rw-r----- 1 jcohen jcohen 169304 Jul  8 13:26 process_data_sub-douglas.log
-rw-r----- 1 jcohen jcohen 245747 Jul  8 13:30 process_data_sub-glen.log
-rw-r----- 1 jcohen jcohen 258671 Jul  8 13:37 process_data_sub-juntendo750w.log
-rw-r----- 1 jcohen jcohen 218961 Jul  8 13:27 process_data_sub-juntendoAchieva.log
-rw-r----- 1 jcohen jcohen 236951 Jul  8 13:31 process_data_sub-juntendoPrisma.log
-rw-r----- 1 jcohen jcohen 242077 Jul  8 13:31 process_data_sub-juntendoSkyra.log
-rw-r----- 1 jcohen jcohen 257784 Jul  8 13:32 process_data_sub-mgh.log
-rw-r----- 1 jcohen jcohen 252596 Jul  8 13:27 process_data_sub-oxfordFmrib.log
-rw-r----- 1 jcohen jcohen 240155 Jul  8 13:32 process_data_sub-perform.log
-rw-r----- 1 jcohen jcohen 250713 Jul  8 13:32 process_data_sub-poly.log
-rw-r----- 1 jcohen jcohen 241182 Jul  8 13:33 process_data_sub-tokyo750w.log
-rw-r----- 1 jcohen jcohen 243668 Jul  8 13:33 process_data_sub-tokyoIngenia.log
-rw-r----- 1 jcohen jcohen 247747 Jul  8 13:35 process_data_sub-tokyoSigna1.log
-rw-r----- 1 jcohen jcohen 230246 Jul  8 13:31 process_data_sub-tokyoSkyra.log
-rw-r----- 1 jcohen jcohen 237806 Jul  8 13:27 process_data_sub-ucl.log
-rw-r----- 1 jcohen jcohen 250507 Jul  8 13:32 process_data_sub-unf.log
-rw-r----- 1 jcohen jcohen   2910 Jul  8 13:38 sct_run_batch_log.txt
~~~

</details>

cfhammill commented:
The printing succeeding subjects was a off by one error in the return code check. That's fixed now. The numeric path is more concerning, because I don't think there is an easy fix. There is a work around, which is just to put the path in quotes, but I'll see if I can find a better solution

jcohenadad commented:
> The numeric path is more concerning, because I don't think there is an easy fix. There is a work around, which is just to put the path in quotes, but I'll see if I can find a better solution

i can see how this is tricky to fix (because we rely on python's yaml parser)-- don't worry about it. Maybe just mention in the doc and examples that path should be with " " (to always avoid the problem)

cfhammill commented:
The difficulty is that there's no easy way to introspect on the parser object to figure out what the types should be, if that was easy then we could just traverse the parsed dictionary and coerce types as necessary. I think it's probably possible but would involve poking around internals in a way that's likely fragile. The more robust solution would be to store the parser arguments as a data frame or dictionary and have a function that produces a parser from that specification and a function that produces a arg/type dictionary.  In either case more work than it's worth given how simple the workaround is. 

I'll add quotes in the help.

cfhammill commented:
Hooray, thanks for all the help @jcohenadad!
